File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/infrastructure/pool/PoolReconciler.kt
/*
 * Copyright 2025 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.sandbox.infrastructure.pool

import com.alibaba.opensandbox.sandbox.domain.pool.PoolConfig
import com.alibaba.opensandbox.sandbox.domain.pool.PoolStateStore
import org.slf4j.LoggerFactory
import java.time.Instant
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService

/**
 * Runs one reconcile tick: leader-gated replenish/shrink and TTL reap.
 *
 * Only the current primary lock holder performs idle maintenance writes.
 * Leader does not voluntarily release the lock; it is only lost when renew fails or TTL expires.
 * Call from a periodic scheduler; [createOne] should call lifecycle create and return the new sandbox ID or null on failure.
 */
internal object PoolReconciler {
    private val logger = LoggerFactory.getLogger(PoolReconciler::class.java)

    /**
     * Runs a single reconcile tick. If this node does not hold the primary lock, returns immediately.
     * Otherwise: reaps expired idle, snapshots counters, then shrinks excess idle or creates up to
     * min(deficit, warmupConcurrency) sandboxes via [createOne] concurrently using [warmupExecutor].
     * Lock is not released at end of tick (distributed implementations rely on TTL or renew failure to release).
     */
    fun runReconcileTick(
        config: PoolConfig,
        stateStore: PoolStateStore,
        createOne: () -> String?,
        onDiscardSandbox: (String) -> Unit = {},
        reconcileState: ReconcileState,
        warmupExecutor: ExecutorService,
    ) {
        val poolName = config.poolName
        val ownerId = config.ownerId
        val ttl = config.primaryLockTtl

        if (!stateStore.tryAcquirePrimaryLock(poolName, ownerId, ttl)) {
            logger.trace("Reconcile skip (not primary): pool_name={}", poolName)
            return
        }
        runPrimaryReplenishOnce(config, stateStore, createOne, onDiscardSandbox, reconcileState, warmupExecutor)
        // Do not release primary lock here; leader holds until renew fails or TTL expires.
    }

    private fun runPrimaryReplenishOnce(
        config: PoolConfig,
        stateStore: PoolStateStore,
        createOne: () -> String?,
        onDiscardSandbox: (String) -> Unit,
        reconcileState: ReconcileState,
        warmupExecutor: ExecutorService,
    ) {
        val poolName = config.poolName
        val ownerId = config.ownerId
        val ttl = config.primaryLockTtl
        val now = Instant.now()

        val discardedAlive = stateStore.reapExpiredIdle(poolName, now, config.acquireMinRemainingTtl)
        for (sandboxId in discardedAlive) {
            // Reaped near-expiry but server-side TTL has not yet elapsed; kill so the live sandbox
            // does not linger past its pool membership and consume quota.
            onDiscardSandbox(sandboxId)
        }
        val counters = stateStore.snapshotCounters(poolName)
        val excess = (counters.idleCount - config.maxIdle).coerceAtLeast(0)
        val toRemove = minOf(excess, config.warmupConcurrency)
        if (toRemove > 0) {
            shrinkExcessIdle(config, stateStore, onDiscardSandbox, toRemove)
            return
        }

        val deficit = (config.maxIdle - counters.idleCount).coerceAtLeast(0)
        val toCreate = minOf(deficit, config.warmupConcurrency)

        if (toCreate == 0 || reconcileState.isBackoffActive(now)) {
            stateStore.renewPrimaryLock(poolName, ownerId, ttl)
            logger.debug(
                "Reconcile tick: pool_name={} idle={} deficit={} toCreate=0 (backoff={})",
                poolName,
                counters.idleCount,
                deficit,
                reconcileState.isBackoffActive(now),
            )
            return
        }

        logger.debug(
            "Reconcile tick: pool_name={} idle={} deficit={} toCreate={}",
            poolName,
            counters.idleCount,
            deficit,
            toCreate,
        )

        if (!stateStore.renewPrimaryLock(poolName, ownerId, ttl)) return
        val tasks = List(toCreate) { Callable { createOne() } }
        val results: List<Pair<String?, String?>> =
            try {
                warmupExecutor.invokeAll(tasks).map { f ->
                    try {
                        f.get() to null
                    } catch (e: Exception) {
                        null to e.message
                    }
                }
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
                logger.info("Reconcile interrupted while waiting warmup tasks: pool_name={}", poolName)
                return
            }

        val createdSandboxIds = mutableListOf<String>()
        var failureCount = 0
        var lastError: String? = null
        for ((newId, errorMessage) in results) {
            if (newId != null) {
                createdSandboxIds += newId
            } else {
                failureCount++
                lastError = errorMessage
            }
        }
        reconcileState.recordFailures(failureCount, lastError)

        var created = 0
        for (index in createdSandboxIds.indices) {
            val sandboxId = createdSandboxIds[index]
            if (!stateStore.renewPrimaryLock(poolName, ownerId, ttl)) {
                val orphanedCount = createdSandboxIds.size - index
                for (orphanedIndex in index until createdSandboxIds.size) {
                    try {
                        onDiscardSandbox(createdSandboxIds[orphanedIndex])
                    } catch (e: Exception) {
                        logger.warn(
                            "Reconcile orphaned sandbox cleanup failed: pool_name={} sandbox_id={} error={}",
                            poolName,
                            createdSandboxIds[orphanedIndex],
                            e.message,
                        )
                    }
                }
                logger.warn(
                    "Reconcile lost primary lock before putIdle; dropped {} newly created sandbox(es): pool_name={}",
                    orphanedCount,
                    poolName,
                )
                return
            }
            try {
                stateStore.putIdle(poolName, sandboxId)
                created++
                reconcileState.recordSuccess()
            } catch (e: Exception) {
                reconcileState.recordFailure(e.message)
                val orphanedCount = createdSandboxIds.size - index
                for (orphanedIndex in index until createdSandboxIds.size) {
                    val orphanedId = createdSandboxIds[orphanedIndex]
                    try {
                        stateStore.removeIdle(poolName, orphanedId)
                    } catch (_: Exception) {
                        // best-effort remove; continue cleanup path
                    }
                    try {
                        onDiscardSandbox(orphanedId)
                    } catch (cleanupError: Exception) {
                        logger.warn(
                            "Reconcile orphaned sandbox cleanup failed after putIdle error: pool_name={} sandbox_id={} error={}",
                            poolName,
                            orphanedId,
                            cleanupError.message,
                        )
                    }
                }
                logger.warn(
                    "Reconcile putIdle failed; dropped {} newly created sandbox(es): pool_name={} error={}",
                    orphanedCount,
                    poolName,
                    e.message,
                )
                return
            }
        }

        if (created > 0) {
            logger.debug("Reconcile created {} sandboxes: pool_name={}", created, poolName)
        }
    }

    private fun shrinkExcessIdle(
        config: PoolConfig,
        stateStore: PoolStateStore,
        onDiscardSandbox: (String) -> Unit,
        toRemove: Int,
    ) {
        val poolName = config.poolName
        val ownerId = config.ownerId
        val ttl = config.primaryLockTtl
        var removed = 0

        repeat(toRemove) {
            if (!stateStore.renewPrimaryLock(poolName, ownerId, ttl)) {
                logger.warn(
                    "Reconcile lost primary lock before shrinking idle: pool_name={} removed={}",
                    poolName,
                    removed,
                )
                return
            }
            val sandboxId = stateStore.tryTakeIdle(poolName) ?: return
            try {
                onDiscardSandbox(sandboxId)
            } catch (e: Exception) {
                logger.warn(
                    "Reconcile shrink sandbox cleanup failed: pool_name={} sandbox_id={} error={}",
                    poolName,
                    sandboxId,
                    e.message,
                )
            }
            removed++
        }

        stateStore.renewPrimaryLock(poolName, ownerId, ttl)
        logger.debug("Reconcile shrunk {} idle sandbox(es): pool_name={}", removed, poolName)
    }
}

File: sdks/sandbox/python/src/opensandbox/_pool_reconciler.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""Sandbox pool reconciliation logic."""

from __future__ import annotations

import logging
from collections.abc import Callable
from concurrent.futures import Executor, wait
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

from opensandbox.pool_types import (
    PoolConfig,
    PoolState,
    PoolStateStore,
)
from opensandbox.pool_types import (
    reap_expired_idle_with_min_ttl as _reap_expired_idle_with_min_ttl,
)

logger = logging.getLogger(__name__)


@dataclass
class ReconcileState:
    degraded_threshold: int
    backoff_base: timedelta = timedelta(seconds=30)
    backoff_max: timedelta = timedelta(days=1)
    failure_count: int = 0
    state: PoolState = PoolState.HEALTHY
    last_error: str | None = None
    backoff_until: datetime | None = None
    backoff_attempts: int = 0

    def record_success(self) -> None:
        self.failure_count = 0
        if self.state == PoolState.DEGRADED:
            self.state = PoolState.HEALTHY
        self.backoff_until = None
        self.backoff_attempts = 0
        self.last_error = None

    def record_failure(self, error_message: str | None) -> None:
        self.record_failures(1, error_message)

    def record_failures(self, count: int, error_message: str | None) -> None:
        if count <= 0:
            return
        self.failure_count += count
        self.last_error = error_message
        if self.failure_count >= self.degraded_threshold:
            self.state = PoolState.DEGRADED
            self.backoff_attempts += 1
            exponent = min(self.backoff_attempts - 1, 30)
            delay = min(
                self.backoff_base.total_seconds() * (1 << exponent),
                self.backoff_max.total_seconds(),
            )
            self.backoff_until = datetime.now(timezone.utc) + timedelta(seconds=delay)

    def is_backoff_active(self, now: datetime | None = None) -> bool:
        until = self.backoff_until
        if until is None:
            return False
        return self.state == PoolState.DEGRADED and (now or datetime.now(timezone.utc)) < until


def run_reconcile_tick(
    *,
    config: PoolConfig,
    state_store: PoolStateStore,
    create_one: Callable[[], str | None],
    on_discard_sandbox: Callable[[str], None],
    reconcile_state: ReconcileState,
    warmup_executor: Executor,
) -> None:
    pool_name = config.pool_name
    owner_id = str(config.owner_id)
    ttl = config.primary_lock_ttl

    if not state_store.try_acquire_primary_lock(pool_name, owner_id, ttl):
        logger.debug("Reconcile skip (not primary): pool_name=%s", pool_name)
        return
    _run_primary_replenish_once(
        config=config,
        state_store=state_store,
        create_one=create_one,
        on_discard_sandbox=on_discard_sandbox,
        reconcile_state=reconcile_state,
        warmup_executor=warmup_executor,
    )


def _run_primary_replenish_once(
    *,
    config: PoolConfig,
    state_store: PoolStateStore,
    create_one: Callable[[], str | None],
    on_discard_sandbox: Callable[[str], None],
    reconcile_state: ReconcileState,
    warmup_executor: Executor,
) -> None:
    pool_name = config.pool_name
    owner_id = str(config.owner_id)
    ttl = config.primary_lock_ttl
    now = datetime.now(timezone.utc)

    discarded_alive = _reap_expired_idle_with_min_ttl(
        state_store, pool_name, now, config.acquire_min_remaining_ttl
    )
    for sandbox_id in discarded_alive:
        # Reaped near-expiry but server-side TTL has not yet elapsed; kill so the live
        # sandbox does not linger past its pool membership and consume quota.
        on_discard_sandbox(sandbox_id)
    counters = state_store.snapshot_counters(pool_name)
    excess = max(0, counters.idle_count - config.max_idle)
    to_remove = min(excess, int(config.warmup_concurrency or 1))
    if to_remove > 0:
        _shrink_excess_idle(config, state_store, on_discard_sandbox, to_remove)
        return

    deficit = max(0, config.max_idle - counters.idle_count)
    to_create = min(deficit, int(config.warmup_concurrency or 1))
    if to_create == 0 or reconcile_state.is_backoff_active(now):
        state_store.renew_primary_lock(pool_name, owner_id, ttl)
        return

    if not state_store.renew_primary_lock(pool_name, owner_id, ttl):
        return

    futures = [warmup_executor.submit(create_one) for _ in range(to_create)]
    wait(futures)
    created_ids: list[str] = []
    failure_count = 0
    last_error: str | None = None
    for future in futures:
        try:
            sandbox_id = future.result()
            if sandbox_id is not None:
                created_ids.append(sandbox_id)
            else:
                failure_count += 1
                last_error = None
        except Exception as exc:
            failure_count += 1
            last_error = str(exc)
    reconcile_state.record_failures(failure_count, last_error)

    created = 0
    for index, sandbox_id in enumerate(created_ids):
        if not state_store.renew_primary_lock(pool_name, owner_id, ttl):
            for orphaned_id in created_ids[index:]:
                _discard(on_discard_sandbox, orphaned_id)
            logger.warning(
                "Reconcile lost primary lock before put_idle; dropped %s newly created sandbox(es): pool_name=%s",
                len(created_ids) - index,
                pool_name,
            )
            return
        try:
            state_store.put_idle(pool_name, sandbox_id)
            created += 1
            reconcile_state.record_success()
        except Exception as exc:
            reconcile_state.record_failure(str(exc))
            for orphaned_id in created_ids[index:]:
                try:
                    state_store.remove_idle(pool_name, orphaned_id)
                except Exception:
                    pass
                _discard(on_discard_sandbox, orphaned_id)
            logger.warning(
                "Reconcile put_idle failed; dropped %s newly created sandbox(es): pool_name=%s error=%s",
                len(created_ids) - index,
                pool_name,
                exc,
            )
            return
    if created > 0:
        logger.debug("Reconcile created %s sandboxes: pool_name=%s", created, pool_name)


def _shrink_excess_idle(
    config: PoolConfig,
    state_store: PoolStateStore,
    on_discard_sandbox: Callable[[str], None],
    to_remove: int,
) -> None:
    pool_name = config.pool_name
    owner_id = str(config.owner_id)
    ttl = config.primary_lock_ttl
    removed = 0
    for _ in range(to_remove):
        if not state_store.renew_primary_lock(pool_name, owner_id, ttl):
            logger.warning(
                "Reconcile lost primary lock before shrinking idle: pool_name=%s removed=%s",
                pool_name,
                removed,
            )
            return
        sandbox_id = state_store.try_take_idle(pool_name)
        if sandbox_id is None:
            return
        _discard(on_discard_sandbox, sandbox_id)
        removed += 1

    state_store.renew_primary_lock(pool_name, owner_id, ttl)
    logger.debug("Reconcile shrunk %s idle sandbox(es): pool_name=%s", removed, pool_name)


def _discard(on_discard_sandbox: Callable[[str], None], sandbox_id: str) -> None:
    try:
        on_discard_sandbox(sandbox_id)
    except Exception as exc:
        logger.warning(
            "Reconcile sandbox cleanup failed: sandbox_id=%s error=%s", sandbox_id, exc
        )
